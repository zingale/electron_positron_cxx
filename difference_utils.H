#ifndef DIFFERENCE_UTILS_H
#define DIFFERENCE_UTILS_H

#include <cassert>
#include <cmath>
#include <functional>
#include <limits>
#include <utility>

// Some high-order finite-difference approximations for the EOS.

// the coefficients for these can be found in many sources, including:
// https://en.wikipedia.org/wiki/Finite_difference_coefficient

template <typename T>
T fourth_order_diff(std::function<T(T)> func, T x0, T delta) {
    // Compute a 4th order accurate centered difference approximation
    // of a function, and allow us to specify the component of the
    // object that is returned (if applicable)

    std::vector<T> fvals;

    for (auto i : {-2, -1, 0, 1, 2}) {
        if (i == 0) {
            fvals.push_back(0.0L);
            continue;
        }

        auto _f = func(x0 + static_cast<T>(i) * delta);
        fvals.push_back(_f);
    }

    T deriv = ((fvals[0] - fvals[4]) +
               8.0L * (-fvals[1] + fvals[3]))
        / (12L * delta);

    return deriv;
}

template <typename T>
T sixth_order_diff(std::function<T(T)> func, T x0, T delta) {
    // Compute a 6th order accurate centered difference approximation
    // of a function, and allow us to specify the component of the
    // object that is returned (if applicable)

    std::vector<T> fvals;

    for (auto i : {-3, -2, -1, 0, 1, 2, 3}) {
        if (i == 0) {
            fvals.push_back(0.0L);
            continue;
        }

        auto _f = func(x0 + static_cast<T>(i) * delta);
        fvals.push_back(_f);
    }

    T deriv = ((-fvals[0] + fvals[6]) +
               9.0 * (fvals[1] - fvals[5]) +
               45.0 * (-fvals[2] + fvals[4])) /
        (60 * delta);

    return deriv;
}

template <typename T>
T eighth_order_diff(std::function<T(T)> func, T x0, T delta) {
    // Compute an 8th order accurate centered difference approximation
    // of a function, and allow us to specify the component of the
    // object that is returned (if applicable)

    std::vector<T> fvals;

    for (auto i : {-4, -3, -2, -1, 0, 1, 2, 3, 4}) {
        if (i == 0) {
            fvals.push_back(0.0L);
            continue;
        }

        auto _f = func(x0 + static_cast<T>(i) * delta);
        fvals.push_back(_f);
    }

    T deriv = ((1.0 / 280.0) * (fvals[0] - fvals[8]) +
               (4.0 / 105.0) * (-fvals[1] + fvals[7]) +
               0.2 * (fvals[2] - fvals[6]) +
               0.8 * (-fvals[3] + fvals[5])) / delta;

    return deriv;
}

template <typename T>
std::pair<T, T>
adaptive_diff(std::function<T(T)> func, T x0, T h, int nlevels=10) {
    // Perform an adaptive centered-difference estimate to the first
    // derivative using Richardson-extrapolation / Ridders' method.

    // see C.J.F. Ridders, "Accurate computation of F'x(x) and F''(x)
    // Advances in Engineering Software, 4, 2, 1978

    // we'll use a second-order centered difference, starting with the
    // initial h.  this has the form:
    //
    //  D_h(f) = 0.5 * (f(x0 + h) - f(x0 - h)) / h + c (h**2) + O(h**4)
    //
    // the key point is that C depends only on the function f (or its
    // derivatives) at point x0.  That means we can compute D_{h/2}(f)
    // and then combine the 2 for a better estimate of the derivative:
    //
    // D' = (4 D_{h/2}(f) - D_h(f)) / 3 + O(h**4)

    // In Ridders' method, we build a tableau of the form:
    //
    //
    //   A(1, 1)  A(1, 2)  A(1, 3)  A(1, 4) ...
    //            A(2, 2)  A(2, 3)  A(2, 4) ...
    //                     A(3, 3)  A(3, 3) ...
    //                              A(4, 4) ...
    //
    // where the entries in each row come via extrapolation of the data
    // in the previous row.
    //
    // In this notation, A(1, 1) = D_h(f),
    //                   A(1, 2) = D_{h/2}(f)
    //                   A(2, 2) = D'
    //
    // we then have:
    //
    // A(1, n) = 0.5 * (f(x + h/2**(n-1)) - f(x - h/2**(n-1))) / (h/2**(n-1))
    //
    // A(m, n) = (4**(m-1) A(m-1, n) - A(m-1, n-1)) / (4**(m-1) - 1)

    // Note: we use 0-based indexing below

    constexpr int max_levels = 50;

    assert (nlevels < max_levels);

    std::array<std::array<T, max_levels>, max_levels> A{};

    assert (h > 0L);

    T _h{h};

    T jump{2.0L};

    auto _fp = func(x0 + _h);
    auto _fm = func(x0 - _h);

    A[0][0] = 0.5L * (_fp - _fm) / _h;
    T deriv = A[0][0];

    auto err = std::numeric_limits<T>::max();

    // loop over columns
    for (int n = 1; n < nlevels; ++n) {
        _h /= jump;
        _fp = func(x0 + _h);
        _fm = func(x0 - _h);

        A[0][n] = 0.5L * (_fp - _fm) / _h;

        // for the current column n, fill in all the rows m > 0
        for (int m = 1; m <= n; ++m) {

            A[m][n] = (std::pow(2.0L * jump, m) * A[m-1][n] - A[m-1][n-1]) /
                (std::pow(2.0L * jump, m) - 1);

            // estimate the error so far, using the current row
            // and the estimates in the row above us
            T err_m = std::max(std::abs(A[m][n] - A[m-1][n]),
                               std::abs(A[m][n] - A[m-1][n-1]));

            if (err_m < err) {
                err = err_m;
                deriv = A[m][n];
            }
        }

        // at this point, A(n, n) should be the most accurate estimate
        // for the derivative, but it's possible that things have
        // become worse -- let's check
        if (std::abs(A[n][n] - A[n-1][n-1]) > 2.0 * err) {
            break;
        }

    }

    return {deriv, err};
}

template <typename T>
std::pair<T, T>
adaptive_diff2(std::function<T(T)> func, T x0, T h, int nlevels=10) {
    // This is like adaptive_diff, but for the second derivative.

    constexpr int max_levels = 50;

    assert (nlevels < max_levels);

    std::array<std::array<T, max_levels>, max_levels> A{};

    assert (h > 0L);

    T _h{h};

    T jump{2.0L};

    auto _fp = func(x0 + _h);
    auto _f0 = func(x0);
    auto _fm = func(x0 - _h);

    A[0][0] = (_fp - 2.0L * _f0 + _fm) / (_h * _h);
    T deriv = A[0][0];

    auto err = std::numeric_limits<T>::max();

    // loop over columns
    for (int n = 1; n < nlevels; ++n) {
        _h /= jump;
        _fp = func(x0 + _h);
        _f0 = func(x0);
        _fm = func(x0 - _h);

        A[0][n] = (_fp - 2.0L * _f0 + _fm) / (_h * _h);

        // for the current column n, fill in all the rows m > 0
        for (int m = 1; m <= n; ++m) {

            A[m][n] = (std::pow(2.0L * jump, m) * A[m-1][n] - A[m-1][n-1]) /
                (std::pow(2.0L * jump, m) - 1);

            // estimate the error so far, using the current row
            // and the estimates in the row above us
            T err_m = std::max(std::abs(A[m][n] - A[m-1][n]),
                               std::abs(A[m][n] - A[m-1][n-1]));

            if (err_m < err) {
                err = err_m;
                deriv = A[m][n];
            }
        }

        // at this point, A(n, n) should be the most accurate estimate
        // for the derivative, but it's possible that things have
        // become worse -- let's check
        if (std::abs(A[n][n] - A[n-1][n-1]) > 2.0 * err) {
            break;
        }

    }

    return {deriv, err};
}


#endif
