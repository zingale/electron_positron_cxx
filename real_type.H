#ifndef REAL_TYPE_H
#define REAL_TYPE_H

#include <string>

#if defined(USE_FLOAT128)
#include <quadmath.h>
using real_t = __float128;
#elif defined(USE_LONG_DOUBLE)
using real_t = long double;
#else
using real_t = double;
#endif

inline namespace literals {

// literal for floating-point constants
// one is for 1.0_rt, and the other is for 1_rt

    constexpr real_t operator""_rt(long double x) {
        return static_cast<real_t>(x);
    }

    constexpr real_t operator""_rt(unsigned long long x) {
        return static_cast<real_t>(x);
    }
}

// std::numbers does not work with __float128

inline
real_t str_to_real_t(const char* s) {
#if defined(USE_FLOAT128)
    return strtoflt128(s, nullptr); 
#elif defined (USE_LONG_DOUBLE)
    return std::stold(s);
#else
    return std::stod(s);
#endif

}

namespace constants {
static inline const real_t pi = str_to_real_t("3.141592653589793238462643383279502884197169399375105820974944592307816406286208998628034825342117068");

static inline const real_t sqrt2 = str_to_real_t("1.414213562373095048801688724209698078569671875376948073176679737990732478462107038850387534327641573");
}

#endif
