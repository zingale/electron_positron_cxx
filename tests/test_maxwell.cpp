#include <print>

#include <array>
#include <cmath>

#include "real_type.H"
#include "electron_positron.H"
#include "difference_utils.H"

constexpr std::array<real_t, 4> Ts{1.e4_rt, 1.e6_rt, 1.e8_rt, 5.e9_rt};
constexpr std::array<real_t, 5> rhos{1.e-2_rt, 1.e2_rt, 1.e5_rt, 1.e7_rt, 5.e9_rt};

// number density

void
test_maxwell() {

    ElectronPositronEOS<real_t> eos;
    constexpr real_t Ye{0.5_rt};

    std::println("testing p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ");

    for (auto T : Ts) {
        for (auto rho : rhos) {
            auto es = eos.pe_state(rho, T, Ye);

            real_t ptot = es.p_e + es.p_pos;
            real_t de_drho = es.dee_drho + es.dep_drho;
            real_t dp_dT = es.dpe_dT + es.dpp_dT;

            real_t term = std::abs((ptot - (rho * rho * de_drho + T * dp_dT)) / ptot);

            std::println("ρ = {:8.3g} T = {:8.3g}:  p⁻ + p⁺ = {:15.8g},  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error = {:15.5g}",
                         rho, T, ptot, term);
        }
    }
}


int main() {

    test_maxwell();

}
