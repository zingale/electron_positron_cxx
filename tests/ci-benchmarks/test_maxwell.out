[92mtesting p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ[0m
ρ =     0.01 T =    1e+04:  p⁻ + p⁺ =   4.6102109e+09,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      5.3827e-34
ρ =      100 T =    1e+04:  p⁻ + p⁺ =    6.807255e+15,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      5.0967e-34
ρ =    1e+05 T =    1e+04:  p⁻ + p⁺ =   6.4992245e+20,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      1.3308e-21
ρ =    1e+07 T =    1e+04:  p⁻ + p⁺ =   8.4599898e+23,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      9.4315e-22
ρ =    5e+09 T =    1e+04:  p⁻ + p⁺ =   4.1972017e+27,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      3.5861e-23
ρ =     0.01 T =    1e+05:  p⁻ + p⁺ =    4.171715e+10,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      1.5863e-34
ρ =      100 T =    1e+05:  p⁻ + p⁺ =   6.8237704e+15,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      3.8133e-34
ρ =    1e+05 T =    1e+05:  p⁻ + p⁺ =   6.4992261e+20,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      6.8131e-28
ρ =    1e+07 T =    1e+05:  p⁻ + p⁺ =   8.4599898e+23,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      8.1172e-21
ρ =    5e+09 T =    1e+05:  p⁻ + p⁺ =   4.1972017e+27,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      3.5918e-22
ρ =     0.01 T =    1e+06:  p⁻ + p⁺ =   4.1576893e+11,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      2.5466e-34
ρ =      100 T =    1e+06:  p⁻ + p⁺ =   8.2955097e+15,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =               0
ρ =    1e+05 T =    1e+06:  p⁻ + p⁺ =   6.4993918e+20,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      1.3994e-33
ρ =    1e+07 T =    1e+06:  p⁻ + p⁺ =   8.4599908e+23,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      1.7103e-24
ρ =    5e+09 T =    1e+06:  p⁻ + p⁺ =   4.1972017e+27,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      3.4264e-21
ρ =     0.01 T =    1e+08:  p⁻ + p⁺ =   4.1572317e+13,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =       4.075e-33
ρ =      100 T =    1e+08:  p⁻ + p⁺ =   4.1616025e+17,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      1.3339e-34
ρ =    1e+05 T =    1e+08:  p⁻ + p⁺ =   7.9828764e+20,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      5.6965e-34
ρ =    1e+07 T =    1e+08:  p⁻ + p⁺ =   8.4696558e+23,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =               0
ρ =    5e+09 T =    1e+08:  p⁻ + p⁺ =   4.1972442e+27,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      1.1361e-34
ρ =     0.01 T =    5e+09:  p⁻ + p⁺ =    2.131083e+24,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =               0
ρ =      100 T =    5e+09:  p⁻ + p⁺ =    2.131083e+24,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      4.3702e-34
ρ =    1e+05 T =    5e+09:  p⁻ + p⁺ =   2.1311937e+24,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      1.0925e-34
ρ =    1e+07 T =    5e+09:  p⁻ + p⁺ =   3.0790866e+24,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      3.0247e-34
ρ =    5e+09 T =    5e+09:  p⁻ + p⁺ =   4.3021496e+27,  p = ρ² ∂e/∂ρ|ᴛ + T ∂p/∂T|ᵨ error =      9.9753e-34

[92mtesting ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ[0m
ρ =     0.01 T =    1e+04:  ∂e/∂T|ᵨ =        59076462,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      7.6573e-34
ρ =      100 T =    1e+04:  ∂e/∂T|ᵨ =       501250.36,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      1.0763e-29
ρ =    1e+05 T =    1e+04:  ∂e/∂T|ᵨ =       5343.9075,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      1.7533e-11
[91mρ =    1e+07 T =    1e+04:  ∂e/∂T|ᵨ =        463.6075,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      2.8387e-09[0m
[91mρ =    5e+09 T =    1e+04:  ∂e/∂T|ᵨ =       50.668913,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      3.1014e-07[0m
ρ =     0.01 T =    1e+05:  ∂e/∂T|ᵨ =        62252584,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      2.0762e-34
ρ =      100 T =    1e+05:  ∂e/∂T|ᵨ =       5003699.5,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      1.5547e-31
ρ =    1e+05 T =    1e+05:  ∂e/∂T|ᵨ =       53439.065,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      4.7705e-20
ρ =    1e+07 T =    1e+05:  ∂e/∂T|ᵨ =        4636.075,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      4.0628e-11
[91mρ =    5e+09 T =    1e+05:  ∂e/∂T|ᵨ =       506.68913,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      3.3462e-09[0m
ρ =     0.01 T =    1e+06:  ∂e/∂T|ᵨ =        62381316,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      5.1797e-34
ρ =      100 T =    1e+06:  ∂e/∂T|ᵨ =        39506027,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      6.5432e-34
ρ =    1e+05 T =    1e+06:  ∂e/∂T|ᵨ =       534381.14,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      7.4286e-30
ρ =    1e+07 T =    1e+06:  ∂e/∂T|ᵨ =       46360.746,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      4.2808e-17
ρ =    5e+09 T =    1e+06:  ∂e/∂T|ᵨ =       5066.8913,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      5.4826e-11
ρ =     0.01 T =    1e+08:  ∂e/∂T|ᵨ =        64921813,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      1.9908e-34
ρ =      100 T =    1e+08:  ∂e/∂T|ᵨ =        64886878,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      3.9838e-34
ρ =    1e+05 T =    1e+08:  ∂e/∂T|ᵨ =        42155312,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =       1.533e-33
ρ =    1e+07 T =    1e+08:  ∂e/∂T|ᵨ =       4632237.3,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      1.6671e-31
ρ =    5e+09 T =    1e+08:  ∂e/∂T|ᵨ =       506684.55,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      9.5477e-29
ρ =     0.01 T =    5e+09:  ∂e/∂T|ᵨ =   6.2562596e+17,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =               0
ρ =      100 T =    5e+09:  ∂e/∂T|ᵨ =   6.2562596e+13,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      1.0831e-34
ρ =    1e+05 T =    5e+09:  ∂e/∂T|ᵨ =   6.2560634e+10,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =               0
ρ =    1e+07 T =    5e+09:  ∂e/∂T|ᵨ =   5.1517971e+08,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =               0
ρ =    5e+09 T =    5e+09:  ∂e/∂T|ᵨ =        24781742,  ∂e/∂T|ᵨ = T ∂s/∂T|ᵨ error =      6.6496e-33

[92mtesting -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ[0m
ρ =     0.01 T =    1e+04:  ∂s/∂ρ|ᴛ =  -3.9384126e+09,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      4.2006e-34
ρ =      100 T =    1e+04:  ∂s/∂ρ|ᴛ =      -3339.3643,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      1.2039e-29
ρ =    1e+05 T =    1e+04:  ∂s/∂ρ|ᴛ =    -0.033464132,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =       1.834e-11
[91mρ =    1e+07 T =    1e+04:  ∂s/∂ρ|ᴛ =  -1.9340336e-05,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =       3.447e-09[0m
[91mρ =    5e+09 T =    1e+04:  ∂s/∂ρ|ᴛ =  -3.3958512e-09,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      3.3115e-07[0m
ρ =     0.01 T =    1e+05:  ∂s/∂ρ|ᴛ =  -4.1499968e+09,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      5.9796e-34
ρ =      100 T =    1e+05:  ∂s/∂ρ|ᴛ =      -33334.964,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      4.1271e-32
ρ =    1e+05 T =    1e+05:  ∂s/∂ρ|ᴛ =     -0.33464126,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =       5.052e-20
ρ =    1e+07 T =    1e+05:  ∂s/∂ρ|ᴛ =  -0.00019340336,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      5.2295e-11
[91mρ =    5e+09 T =    1e+05:  ∂s/∂ρ|ᴛ =    -3.39585e-08,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      3.5888e-09[0m
ρ =     0.01 T =    1e+06:  ∂s/∂ρ|ᴛ =  -4.1570021e+09,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      1.6914e-33
ρ =      100 T =    1e+06:  ∂s/∂ρ|ᴛ =      -263160.62,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      9.5924e-34
ρ =    1e+05 T =    1e+06:  ∂s/∂ρ|ᴛ =      -3.3463509,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      4.8083e-29
ρ =    1e+07 T =    1e+06:  ∂s/∂ρ|ᴛ =   -0.0019340334,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      5.8841e-17
ρ =    5e+09 T =    1e+06:  ∂s/∂ρ|ᴛ =    -3.39585e-07,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      6.0203e-11
ρ =     0.01 T =    1e+08:  ∂s/∂ρ|ᴛ =  -4.1572311e+09,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      4.5565e-32
ρ =      100 T =    1e+08:  ∂s/∂ρ|ᴛ =      -415484.33,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =       8.506e-34
ρ =    1e+05 T =    1e+08:  ∂s/∂ρ|ᴛ =      -261.29094,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      3.7739e-34
ρ =    1e+07 T =    1e+08:  ∂s/∂ρ|ᴛ =      -0.1932346,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      4.7168e-31
ρ =    5e+09 T =    1e+08:  ∂s/∂ρ|ᴛ =  -3.3958193e-05,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      1.3802e-28
ρ =     0.01 T =    5e+09:  ∂s/∂ρ|ᴛ =  -1.9024772e+19,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =               0
ρ =      100 T =    5e+09:  ∂s/∂ρ|ᴛ =  -1.9024772e+11,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =               0
ρ =    1e+05 T =    5e+09:  ∂s/∂ρ|ᴛ =      -190242.41,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =               0
ρ =    1e+07 T =    5e+09:  ∂s/∂ρ|ᴛ =       -16.11964,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      1.9116e-34
ρ =    5e+09 T =    5e+09:  ∂s/∂ρ|ᴛ =   -0.0016608482,  -∂s/∂ρ|ᴛ = 1/ρ² ∂p/∂T|ᵨ error =      9.1727e-33
